### PR TITLE
Add change log with towncrier

### PR DESCRIPTION
**Pull request by decentral1se**
Not working yet but we can probably bikeshed on stuff already.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -17,3 +17,4 @@ pytestdebug.log
 ubuntu-xenial-16.04-cloudimg-console.log
 .python-version
 coverage.xml
+pip-wheel-metadata
```
Comment (webknjaz) [inline .gitignore]: I believe that this should probably go to a separate PR. (Personally, though, I'd prefer having it based on the gitignore.io templates as it's easier to keep up with standards this way...)

File: tox.ini [hunk 2]
```diff
@@ -126,7 +126,6 @@ commands =
 whitelist_externals =
     sh
 
-
```
Comment (webknjaz) [inline tox.ini]: (unrelated change)

File: tox.ini [hunk 3]
```diff
@@ -159,3 +158,12 @@ commands =
 whitelist_externals =
     rm
     {[testenv]whitelist_externals}
+
+[testenv:changelog]
+skip_install = false
+usedevelop = false
+description = Generate the changelog
+extras =
+  changelog
```
Comment (webknjaz) [inline tox.ini]: I believe that for towncrier to work installing dist is not required. So, if you remove that, this task will complete faster.

Currently, there's no way to tell pip "install extras but not molecule itself" which means that it should probably be left out of dist metadata.

File: tox.ini [hunk 4]
```diff
@@ -159,3 +158,12 @@ commands =
...
+commands =
+  towncrier
```
Comment (webknjaz) [inline tox.ini]: Please use four spaces to follow the style in other envs.

File: setup.cfg [hunk 5]
```diff
@@ -112,6 +112,8 @@ vagrant =
     python-vagrant
 windows =
     pywinrm
+changelog =
```
Comment (webknjaz) [inline setup.cfg]: https://github.com/ansible/molecule/pull/1963/files#r274377937

File: setup.cfg [hunk 6]
```diff
@@ -112,6 +112,8 @@ vagrant =
...
+    towncrier
```
Comment (webknjaz) [inline setup.cfg]: Also, take into account that it's not a runtime dependency meaning that there's no benefit of exposing this to users.

I didn't state it anywhere in Molecule forums, _I think_, but I usually consider extras as a public API and try to apply other public APIs considerations to extras.

File: changelog/_template.rst [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+{% if sections[section] %}
+{% for category, val in definitions.items() if category in sections[section] %}
+
+{{ definitions[category]['name'] }}
+{{ underline * definitions[category]['name']|length }}
+
+{% if definitions[category]['showcontent'] %}
+{% for text, values in sections[section][category]|dictsort(by='value') %}
+{% set issue_joiner = joiner(', ') %}
+- {% for value in values|sort %}{{ issue_joiner() }}`{{ value }} <https://github.com/ansible/molecule/issues/{{ value[1:] }}>`_{% endfor %}: {{ text }}
```
Comment (webknjaz) [inline changelog/_template.rst]: I think, this should use extlinks declared in the sphinx config — https://github.com/ansible/molecule/blob/6b6b2bd/doc/source/conf.py#L72-L75
This is how:
```suggestion
- {% for value in values|sort %}{{ issue_joiner() }}:issue:`{{ value }} <{{ value[1:] }}`_{% endfor %}: {{ text }}
```
There's: `:issue:`, `:pr:`, `:commit:` and `:gh:` defined.
`:gh:` one is already used in the changelog, for example: https://github.com/ansible/molecule/blame/master/CHANGELOG.rst#L81

File: changelog/_template.rst [hunk 8]
```diff
@@ -0,0 +1,40 @@
+{% for section in sections %}
+{% set underline = "-" %}
```
Comment (webknjaz) [inline changelog/_template.rst]: I think we should use indentation in this jinja template for better readability. For HTML-like source, it's usually 2 spaces.

File: .gitignore (hunk 1 above)
Comment (decentral1se) [reply .gitignore]: See https://github.com/ansible/molecule/pull/1978.

**Comment by decentral1se** [comment]
Uhhh!? Bizarre ... I forced pushed to remove unrelated changes and it is now closed and the "re-open" button cannot be clicked .... well, I'll open a new PR when I get time then. Thanks for review on this.

**Comment by webknjaz** [comment]
Ah.. Happens :)
